src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,481 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from enum import Enum
+
+# NB: Avoid using "Optional" and use "| None" instead, as "Option" and "Optional" look similar.
+from typing import TYPE_CHECKING, Any, Dict, Generic, Type, TypeVar, cast, overload
+
+from pants.option import custom_types
+from pants.util.frozendict import FrozenDict
+
+if TYPE_CHECKING:
+    from pants.option.subsystem import Subsystem
+
+SubsystemT = TypeVar("SubsystemT", bound="Subsystem")
+OptionT = TypeVar("OptionT", bound="OptionBase")
+PropType = TypeVar("PropType")
+StrT = TypeVar("StrT", str, "str | None")
+IntT = TypeVar("IntT", int, "int | None")
+FloatT = TypeVar("FloatT", float, "float | None")
+BoolT = TypeVar("BoolT", bool, "bool | None")
+EnumT = TypeVar("EnumT", bound=Enum)
+ValueType = TypeVar("ValueType")
+# NB: We don't provide constraints, as our `XListOption` types act like a set of contraints
+ListMemberType = TypeVar("ListMemberType")
```

Comment:
Maybe make these private with `_` so that they don't show up with IDE auto-complete? They're not intended to be consumed externally.

Reply:
I'll follow Pants' style here. But My preference is keep them public because they show up in error messages, and the underscore is annoying.
But then we'd have to define `__all__` or similar. 

There's no winning

Reply:
Will they show up in error messages? MyPy ones maybe?

Reply:
Yup!